### PR TITLE
Feature/timeout

### PR DESCRIPTION
**Pull request by aewag**
Redesign of #79

### CODE REVIEW

**Comment by aewag** [comment]
With this timeout implementation the end_reason is not set, as the qemu process is terminated early.

```
$ h5dump output.hdf5 | less | grep -A10 end_reason
               ATTRIBUTE "end_reason" {
                  DATATYPE  H5T_STRING {
                     STRSIZE 12;          
                     STRPAD H5T_STR_NULLTERM;
                     CSET H5T_CSET_UTF8;
                     CTYPE H5T_C_S1;                
                  }                                 
                  DATASPACE  SCALAR                 
                  DATA {                            
                  (0): "not executed"               
                  }                                 
--                                                  
               ATTRIBUTE "end_reason" {
                  DATATYPE  H5T_STRING {
                     STRSIZE 12;          
                     STRPAD H5T_STR_NULLTERM;
                     CSET H5T_CSET_UTF8;
                     CTYPE H5T_C_S1;                
                  }                                 
                  DATASPACE  SCALAR                 
                  DATA {                            
                  (0): "not executed"               
                  }                                 
--                                                  
               ATTRIBUTE "end_reason" {
                  DATATYPE  H5T_STRING {
                     STRSIZE 12;          
                     STRPAD H5T_STR_NULLTERM;
                     CSET H5T_CSET_UTF8;
                     CTYPE H5T_C_S1;                
                  }                                 
                  DATASPACE  SCALAR                 
                  DATA {                            
                  (0): "not executed"               
                  }
--
         ATTRIBUTE "end_reason" {
            DATATYPE  H5T_STRING {
               STRSIZE 20;
               STRPAD H5T_STR_NULLTERM;
               CSET H5T_CSET_UTF8;
               CTYPE H5T_C_S1;
            }
            DATASPACE  SCALAR
            DATA {
            (0): "endpoint 134217814/1"
            }
--
         ATTRIBUTE "end_reason" {
            DATATYPE  H5T_STRING {
               STRSIZE 20;
               STRPAD H5T_STR_NULLTERM;
               CSET H5T_CSET_UTF8;
               CTYPE H5T_C_S1;
            }
            DATASPACE  SCALAR
            DATA {
            (0): "endpoint 134217834/1"
            }
--
            ATTRIBUTE "end_reason" {
               DATATYPE  H5T_STRING {
                  STRSIZE 1;
                  STRPAD H5T_STR_NULLTERM;
                  CSET H5T_CSET_UTF8;
                  CTYPE H5T_C_S1;
               }
               DATASPACE  NULL
               DATA {
               }
            }
--
            ATTRIBUTE "end_reason" {
               DATATYPE  H5T_STRING {
                  STRSIZE 1;
                  STRPAD H5T_STR_NULLTERM;
                  CSET H5T_CSET_UTF8;
                  CTYPE H5T_C_S1;
               }
               DATASPACE  NULL
               DATA {
               }
            }
--
            ATTRIBUTE "end_reason" {
               DATATYPE  H5T_STRING {
                  STRSIZE 1;
                  STRPAD H5T_STR_NULLTERM;
                  CSET H5T_CSET_UTF8;
                  CTYPE H5T_C_S1;
               }
               DATASPACE  NULL
               DATA {
               }
            }
```

File: fault-readme.md [hunk 1]
```diff
@@ -233,4 +233,7 @@ Use of the ring buffer implementation to store the list of executed translation
 Enable collection of data on all memory accesses. The configuration property expects to be passed a boolean value. If unspecified, it will default to `false`.
 
 ### timeout
-Maximum execution duration in seconds for a single experiment. If exceeded, ARCHIE will be stopped. If unspecified, it will default to `1200` seconds.
+Maximum execution duration in seconds for a single experiment.
+If exceeded, ARCHIE will be stopped.
```
Comment (lukasauer) [inline fault-readme.md:237]: Should be "the experiment will be stopped."

File: controller.py [hunk 2]
```diff
@@ -826,7 +826,7 @@ def controller(
                 clogger.debug(f"Process time running mean: {p_time_mean:.3f}s")
                 # Remove process from list
                 p_list.pop(i)
-                break
+                break  # TODO AW: why break here and not iterate over all processes?
```
Comment (lukasauer) [inline controller.py:818]: I guess the idea was to immediately schedule a new task once a process is freed up. I am not sure though if that makes sense. I would also iterate over all processes here.

File: controller.py [hunk 3]
```diff
@@ -830,29 +819,22 @@ def controller(
...
-    if faultlist:
-        tperindex = (t1 - t0) / len(faultlist)
-    else:
-        tperindex = t1 - t0
-
-    tperworker = tperindex / num_workers
+    time_per_experiment = total_runtime / len(faultlist) if faultlist else total_runtime
+    time_per_worker = (
+        time_per_experiment / num_workers
+    )  # TODO AW: why divide by the number of workers?
```
Comment (lukasauer) [inline controller.py:831]: Not sure how useful that statistic is. We could also drop it.

**Review by lukasauer — APPROVED** [review]
Except for the todo messages, LGTM.

File: controller.py (hunk 3 above)
Comment (aewag) [reply controller.py:831]: Dropped.

File: controller.py (hunk 2 above)
Comment (aewag) [reply controller.py]: :+1: Sounds good, just fixed and tested it with the example.

**Pull request merged**